enclave/core/sgx/backtrace.c

```diff
@@ -83,3 +85,99 @@ int oe_backtrace(void** buffer, int size)
     return 0;
 #endif
 }
+
+char** oe_backtrace_symbols(void* const* buffer, int size)
+{
+    /* Backtrace must use the internal allocator to bypass debug-malloc. */
+    extern void* dlmalloc(size_t size);
+    extern void dlfree(void* ptr);
+    extern void* dlrealloc(void* ptr, size_t size);
+    char** ret = NULL;
+    void* buf = NULL;
+    const size_t BUF_SIZE = 4096;
+    size_t buf_size = BUF_SIZE;
+    size_t buf_size_out;
+    uint32_t retval;
+    char** argv = NULL;
+
+    if (!buffer || size < 0)
+        goto done;
+
+    if (!(buf = dlmalloc(buf_size)))
+        goto done;
+
+    /* First call might return OE_BUFFER_TOO_SMALL. */
+    if (oe_backtrace_symbols_ocall(
+            &retval,
+            oe_get_enclave(),
+            (const uint64_t*)buffer,
+            (size_t)size,
+            buf,
+            buf_size,
+            &buf_size_out) != OE_OK)
+    {
+        goto done;
+    }
+
+    /* Second call uses buffer size returned by first call. */
+    if ((oe_result_t)retval == OE_BUFFER_TOO_SMALL)
+    {
+        buf_size = buf_size_out;
+
+        if (!(buf = dlrealloc(buf, buf_size)))
+            goto done;
+
+        if (oe_backtrace_symbols_ocall(
+                &retval,
+                oe_get_enclave(),
+                (const uint64_t*)buffer,
+                (size_t)size,
+                buf,
+                buf_size,
+                &buf_size_out) != OE_OK)
+        {
+            goto done;
+        }
+
+        if ((oe_result_t)retval != OE_OK)
+            goto done;
+    }
+    else if ((oe_result_t)retval != OE_OK)
+    {
+        goto done;
+    }
+
+    /* Convert vector to array of strings. */
+    {
+        oe_vector_t* vec;
+
+        if (!(vec = oe_vector_relocate(buf, (size_t)size)))
```

Comment:
It is not clear why we need a new concept of a vector. This vector does not seem to have the same semantics as std::vector.
What are the security guarantees of this vector if the host intentionally manipulates the vector's bytes?

Consider just writing strings one after the other onto a buffer and then reading them one after the other at the receiver with proper null checks.

Reply:
I went with your suggested approach and backed out the vector type. We can discuss later, but some utilities are needed to compenstate for the lack of various features in EDL. I will expound on this in my memory model paper.